### PR TITLE
tests: Adds configurable docker conf for test

### LINKED ISSUE #84609: tests: should be able to pull from private registry with secret on Windows containerd
**Which jobs are failing**:

containerd-l2bridge-windows-master
containerd-overlay-windows-master

**Which test(s) are failing**:

[k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull from private registry with secret [NodeConformance]

**Since when has it been failing**:

Unknown.

**Testgrid link**:

https://testgrid.k8s.io/sig-windows#containerd-l2bridge-windows-master
https://testgrid.k8s.io/sig-windows#containerd-overlay-windows-master

**Reason for failure**:

When creating the container, the following error occurs:

    Failed: Error: failed to create containerd task: hcsshim::CreateComputeSystem image-pull-test: The container operating system does not match the host operating system.

Windows Server 2019 is used in the testing environment, and the image ``gcr.io/authenticated-image-pulling/windows-nanoserver:v1`` is supposed to be based on the ``windows/nanoserver:1809`` image, and the test is passing in all other jobs that are using docker instead of containerd.

But the image ``gcr.io/authenticated-image-pulling/windows-nanoserver:v1`` has been built at the beggining of 2019, so it's most likely outdated (the Microsoft official images are updated regularly), which could lead to the OS version mismatch.

Note that most of the other tests are passing for the containerd jobs, which are using the test images hosted in the ``e2eteam`` registry, which are based on the ``mcr.microsoft.com/windows/servercore:ltsc2019`` image (LTS 2019 and 1809 have the same OS version), and they're always up-to-date.

One solution could be to republish the ``gcr.io/authenticated-image-pulling/windows-nanoserver:v1`` image based on a freshly pulled nanoserver image.

**Anything else we need to know**:

/sig windows
/sig testing

### LINKED ISSUE #87785: Windows 1909 failing test: E2E: [k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull from private registry with secret [NodeConformance]
"[k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull from private registry with secret [NodeConformance]" is failing consistently for Windows 1909 on GCE, but it's passing for 2019. Likely one of the containers it depends on hasn't been rebuilt as multi-arch?

https://testgrid.k8s.io/sig-windows#gce-windows-1909-master&width=20

Failure message from https://storage.googleapis.com/kubernetes-jenkins/logs/ci-kubernetes-e2e-windows-gce-1909/1224386533100883969/build-log.txt:
```
[91m[1m• Failure [932.789 seconds][0m
[k8s.io] Container Runtime
[90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/framework/framework.go:697[0m
  blackbox test
  [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/runtime.go:38[0m
    when running a container with a new image
    [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/runtime.go:263[0m
      [91m[1mshould be able to pull from private registry with secret [NodeConformance] [It][0m
      [90m/go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/runtime.go:385[0m

      [91mFeb  3 18:44:35.174: All 3 attempts failed: expected container state: Running, got: "Terminated"[0m

      /go/src/k8s.io/kubernetes/_output/dockerized/go/src/k8s.io/kubernetes/test/e2e/common/runtime.go:364
[90m------------------------------[0m
{"msg":"FAILED [k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull from private registry with secret [NodeConformance]","total":-1,"completed":13,"skipped":194,"failed":1,"failures":["[k8s.io] Container Runtime blackbox test when running a container with a new image should be able to pull from private registry with secret [NodeConformance]"]}
```

### PR DESCRIPTION
**Pull request by claudiubelu**
**What type of PR is this?**

/kind failing-test

/sig windows
/sig testing

**What this PR does / why we need it**:

The image ``gcr.io/authenticated-image-pulling/windows-nanoserver:v1`` is not a manifest list, and it is only useful for Windows Server 1809, which means that the test ``should be able to pull from private registry with secret`` will fail for environments with Windows Server 1903, 1909, or any other future version we might want to test.

This PR adds the the ability to have an alternative private image to pull by using a configurable docker config file which contains the necessary credentials needed to pull the image.

**Which issue(s) this PR fixes**:

Fixes #84609
Fixes #87785

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: test/e2e/common/runtime.go [hunk 1]
```diff
@@ -294,7 +295,12 @@ while true; do sleep 1; done
 		}
 	}
 }`
-
+					// we might be told to use a different docker config JSON.
+					if framework.TestContext.DockerConfigFile != "" {
+						contents, err := ioutil.ReadFile(framework.TestContext.DockerConfigFile)
+						framework.ExpectNoError(err)
```
Comment (neolit123) [inline test/e2e/common/runtime.go]: i'd add an annotation here explaining the error in more detail:
`...err, "could not read the Docker configuration file")`

File: test/e2e/framework/test_context.go [hunk 2]
```diff
@@ -302,6 +305,7 @@ func RegisterCommonFlags(flags *flag.FlagSet) {
 
 	flags.StringVar(&TestContext.ProgressReportURL, "progress-report-url", "", "The URL to POST progress updates to as the suite runs to assist in aiding integrations. If empty, no messages sent.")
 	flags.StringVar(&TestContext.SpecSummaryOutput, "spec-dump", "", "The file to dump all ginkgo.SpecSummary to after tests run. If empty, no objects are saved/printed.")
+	flags.StringVar(&TestContext.DockerConfigFile, "docker-config-file", "", "A file that contains credentials which can be used to pull images from certain private registries, needed for a test.")
```
Comment (neolit123) [inline test/e2e/framework/test_context.go]: i'd reword to:
> A file that contains Docker credentials which can be used to pull test images from private registries

**Comment by neolit123** [comment]
i'm overall +1 on this addition. added a couple of minor comments.

/approve
needs a /test + /hack appovers.

**Comment by PatrickLang** [comment]
/lgtm
This will help test Windows previews as well which may not be available on public repos. Everything should be a parameter so we can override in different test environments. It also doesn't make sense to put every prerelease build in a public repo because they're not broadly used.

**Comment by PatrickLang** [comment]
/approve
(probably won't work but why not)

**Comment by oomichi** [comment]
@claudiubelu
Please check @neolit123 comments.
The other part looks good for me.

/lgtm
/approve

**Comment by BenTheElder** [comment]
I still think our tests should be usable without the mess of wrappers we keep adding to the binary, but that's a battle for another day.
/approve
for hack/ changes specifically, these follow the existing patterns, I did NOT review the rest.
